### PR TITLE
Feature/groups - first draft WIP

### PR DESCRIPTION
**Pull request by mynhiennguyen**
**What was done?**
- addition of a simple select that allows the user to swap between groups (=canvases)
- adding group-info to all message types
- add new Action (`SendHistoryAction`) and Message (`changeGroup`) types to FE + BE
- rework BE to support Groups

**Next steps**
- let users dynamically add / delete / edit groups
- let users invite other users to groups... -> 'login' / authentication / identification needed?

### CODE REVIEW

File: be/server.ts [hunk 1]
```diff
@@ -1,24 +1,29 @@
...
-import { updateUndoRedoAvailabilities } from "./src/utils/utils";
 import  WebSocket  from "ws";
 import express from 'express';
+import { Group } from "./src/types/group";
 
 const PORT = process.env.PORT || 3000;
 
+// init 3 groups
+// TODO: dynamic groups
+const groups = new Map<string, Group>([["Group A", new Group("Group A")], ["Group B", new Group("Group B")], ["Group C", new Group("Group C")]]);
```
Comment (mynhiennguyen) [inline be/server.ts:12]: currently hardcoded 3 groups - goal will be to have the user create / delete / edit their own groups

File: be/server.ts [hunk 2]
```diff
@@ -29,8 +34,19 @@ wss.on("connection", (ws: WebSocket) => {;
   // handle incoming messages
   ws.on("message", (msg: string) => {
     const action: Action = MessageDecoder.parse(msg);
-    action.pushTo(history);
-    broadcastToClients(action.createMessage(ws)); // TODO: remove ws as parameter if possible
+
+    if(action instanceof SendHistoryAction){
```
Comment (mynhiennguyen) [inline be/server.ts:38]: this check seems odd 😓

**Comment by mynhiennguyen** [comment]
**Question:** in order to have users invite other users to groups / manage group - some sort of login / authentication is needed? Any advice on how to approach this?

File: be/server.ts [hunk 3]
```diff
@@ -46,36 +62,35 @@ const handleNewClientConnection = ((ws: WebSocket) => {
   // add new user to list of active users and broadcast list to all other users
   activeUsers.set(ws, newUser);
   broadcastListOfActiveUsers(activeUsers);
-  // send current history to client
-  // TODO: create message out of this?
-  history.undoStack.forEach((m: Action) => {
+  // send history of default group (Group A) to client
+  groups.get("Group A")?.historyStack.undoStack.forEach((m: Action) => {
```
Comment (Garbanas) [inline be/server.ts:66]: Do not use "magic" strings (or numbers). When "Group A" is your default, denn add a constant somewhere like
```
const DEFAULT_GROUP_KEY = "Group A";
...
groups.get(DEFAULT_GROUP_KEY)?....
```

File: be/src/types/action.ts [hunk 4]
```diff
@@ -16,13 +17,14 @@ export abstract class Action<
 > {
   constructor(
     public readonly type: MessageType,
+    public readonly group: string,
```
Comment (Garbanas) [inline be/src/types/action.ts:20]: When you have a class "Group" with property "id: string", then better rename the parameter that has the group.id from "group" to "groupId" to be more specific. What if you wanted to pass the "Group" class?

File: be/src/types/group.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import { HistoryStack } from "./history";
+import { updateUndoRedoAvailabilities } from "../utils/utils";
+
+/**
+ * A group owns its own canvas + historyStack.
+ */
+export class Group {
+    historyStack: HistoryStack;
+
+    constructor(public readonly id: String) {
```
Comment (Garbanas) [inline be/src/types/group.ts:10]: `public readonly id: String` can be `public readonly id: string`.
"String" is a class / object and "string" is a primitive

File: be/src/types/group.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import { HistoryStack } from "./history";
+import { updateUndoRedoAvailabilities } from "../utils/utils";
+
+/**
+ * A group owns its own canvas + historyStack.
+ */
+export class Group {
```
Comment (Garbanas) [inline be/src/types/group.ts:7]: A group could have an ID (for internal use, sending as identifier) and a (display-)name (visible for humans).
So you could change the name of the group without affecting any history stack or message sending.

**Review by Garbanas — CHANGES_REQUESTED** [review]

**Pull request merged**